### DISCUSSION TITLE
Move a window to the next lowest numbered empty workspace

### DISCUSSION THREAD

**Discussion by gmayerw** [xmonad/xmonad#369] [Q&A] [answered]
How would I move the currently selected window on any workspace to the lowest numbered empty workspace? Can it also switch to the workspace that the window is moved to?

For example,
-I have msn.com on workspace 1
-I have a tall screen on workspace 2 with 2 windows open:
    -Code
    -A reference sheet

Let's say I run the the desired code on a new instance of my browser while I am on workspace 5. The desired behavior is that xmonad moves my browser window to workspace 3 because 1 and 2 are occupied.

Now I close msn.com on workspace 1 so that all I have is:
-Workspace 1 is empty
-I have a tall screen on workspace 2 with 2 windows open:
    -Code
    -A reference sheet
-I have that same browser window on workspace 3

Assuming I am on workspace 3 currently and I hit the shortcut key, my browser window and me should immediately move to workspace 1 because it is lower than 3 and it is currently empty.

Let me know if this is clear or not. Thanks!

**Comment by geekosaur** [comment, marked as answer]
[XMonad.Actions.CycleWS](https://hackage.haskell.org/package/xmonad-contrib-0.17.0/docs/XMonad-Actions-CycleWS.html) supports much of what you want, although it would select the first empty workspace in the direction you specify (so if you have two empty workspaces before workspace 5 it would select the second instead of the first). The operations it provides only do one or the other of moving the window or `view`ing the workspace, but it also provides a more general function that would let you do both at once.

The thing you'd bind to your chosen key would look something like
```hs
doTo Prev emptyWS (\ws -> withFocused (\w -> windows (W.view ws <> W.shiftWin ws w)))
```
Note that I have not tested this and it may not be quite what you want.

**Reply by gmayerw** [reply]
`xmonad.hs`:
```
{-# LANGUAGE NoMonomorphismRestriction #-}
import XMonad hiding ((|||))
import XMonad.Util.EZConfig (additionalKeysP)
import XMonad.Layout.LayoutCombinators
import qualified XMonad.StackSet as W
import XMonad.Actions.GridSelect
import qualified Data.Map as M
import XMonad.Actions.CycleWS


myNavigation :: TwoD a (Maybe a)
myNavigation = makeXEventhandler $ shadowWithKeymap navKeyMap navDefaultHandler
 where navKeyMap = M.fromList [
          ((0,xK_Escape), cancel)
         ,((0,xK_space) , select)
         ,((0,xK_slash) , substringSearch myNavigation)
         ,((0,xK_Left)  , move (-1,0)  >> myNavigation)
         ,((0,xK_h)     , move (-1,0)  >> myNavigation)
         ,((0,xK_Right) , move (1,0)   >> myNavigation)
         ,((0,xK_l)     , move (1,0)   >> myNavigation)
         ,((0,xK_Down)  , move (0,1)   >> myNavigation)
         ,((0,xK_j)     , move (0,1)   >> myNavigation)
         ,((0,xK_Up)    , move (0,-1)  >> myNavigation)
         ,((0,xK_k)     , move (0,-1)  >> myNavigation)
         ,((0,xK_y)     , move (-1,-1) >> myNavigation)
         ,((0,xK_i)     , move (1,-1)  >> myNavigation)
         ,((0,xK_n)     , move (-1,1)  >> myNavigation)
         ,((0,xK_m)     , move (1,-1)  >> myNavigation)
         ,((0,xK_Return), setPos (0,0) >> myNavigation)
         ]
       navDefaultHandler = const myNavigation

myColorizer =
  colorRangeFromClassName
    (0x00, 0x00, 0x00) -- lowest inactive bg
    (0x00, 0x00, 0x00) -- highest inactive bg
    (0xff, 0xff, 0xff) -- active bg
    (0xff, 0xff, 0xff) -- inactive fg
    (0x00, 0x00, 0x00) -- active fg

gsconfig1 colorizer =
  (buildDefaultGSConfig myColorizer)
    { gs_cellheight = 50
    , gs_cellwidth = 300
    , gs_cellpadding = 10
    , gs_bordercolor = "black"
    , gs_font = "xft:DejaVuSans-24:regular"
    , gs_navigate = myNavigation
    }

main :: IO ()
main = xmonad $ def
        { modMask = mod4Mask
        , terminal = "xterm -e vifm -c empty /home/wgm"
        , borderWidth = 0
        , layoutHook = Full ||| tiled
        }
      `additionalKeysP`
        [ ("M-S-<Space>", gets windowset >>= mapM_ (windows . W.shiftWin "1") . W.allWindows >> windows (W.view "1") >> sendMessage (JumpToLayout "Tall"))
        , ("M-;", goToSelected $ gsconfig1 myColorizer)
        , ("M-C-<Space>", doTo Prev emptyWS (\ws -> withFocused (\w -> windows (W.view ws <> W.shiftWin ws w))))
        ]

tiled = Tall 1 (3/100) (1/2)
```

Output:
```
XMonad will use ghc to recompile, because "/home/wgm/.xmonad/build" does not exist.
Error detected while loading xmonad configuration file: /home/wgm/.xmonad/xmonad.hs

xmonad.hs:61:37: error:
    • Variable not in scope: emptyWS :: WSType
    • Perhaps you meant one of these:
        data constructor ‘EmptyWS’ (imported from XMonad.Actions.CycleWS),
        ‘M.empty’ (imported from Data.Map)
   |
61 |         , ("M-C-<Space>", doTo Prev emptyWS (\ws -> withFocused (\w -> windows (W.view ws <> W.shiftWin ws w))))
   |                                     ^^^^^^^

Please check the file for errors.

xmonad: xmessage: executeFile: does not exist (No such file or directory)


```

**Reply by geekosaur** [reply]
Sigh, that's right, you're still on 0.15. You want `EmptyWS` instead of `emptyWS`.

**Reply by gmayerw** [reply]
```
XMonad will use ghc to recompile, because "/home/wgm/.xmonad/build" does not exist.
Error detected while loading xmonad configuration file: /home/wgm/.xmonad/xmonad.hs

xmonad.hs:61:27: error:
    • Couldn't match expected type: X ()
                  with actual type: (WorkspaceId -> X ()) -> X ()
    • Probable cause: ‘doTo’ is applied to too few arguments
      In the expression:
        doTo
          Prev EmptyWS
          (\ ws
             -> withFocused (\ w -> windows (W.view ws <> W.shiftWin ws w)))
      In the expression:
        ("M-C-<Space>",
         doTo
           Prev EmptyWS
           (\ ws
              -> withFocused (\ w -> windows (W.view ws <> W.shiftWin ws w))))
      In the second argument of ‘additionalKeysP’, namely
        ‘[("M-S-<Space>",
           gets windowset >>= mapM_ (windows . W.shiftWin "1") . W.allWindows
             >> windows (W.view "1")
             >> sendMessage (JumpToLayout "Tall")),
          ("M-;", goToSelected $ gsconfig1 myColorizer),
          ("M-C-<Space>",
           doTo
             Prev EmptyWS
             (\ ws
                -> withFocused (\ w -> windows (W.view ws <> W.shiftWin ws w))))]’
   |
61 |         , ("M-C-<Space>", doTo Prev EmptyWS (\ws -> withFocused (\w -> windows (W.view ws <> W.shiftWin ws w))))
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

xmonad.hs:61:46: error:
    • Couldn't match expected type: X XMonad.Util.WorkspaceCompare.WorkspaceSort
                  with actual type: WorkspaceId -> X ()
    • The lambda expression ‘\ ws -> withFocused (\ w -> ...)’
      has one value argument,
        but its type ‘X XMonad.Util.WorkspaceCompare.WorkspaceSort’
        has none
      In the third argument of ‘doTo’, namely
        ‘(\ ws
            -> withFocused (\ w -> windows (W.view ws <> W.shiftWin ws w)))’
      In the expression:
        doTo
          Prev EmptyWS
          (\ ws
             -> withFocused (\ w -> windows (W.view ws <> W.shiftWin ws w)))
   |
61 |         , ("M-C-<Space>", doTo Prev EmptyWS (\ws -> withFocused (\w -> windows (W.view ws <> W.shiftWin ws w))))
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Please check the file for errors.

xmonad: xmessage: executeFile: does not exist (No such file or directory)
```

**Reply by geekosaur** [reply]
Argh. I am indeed missing a parameter. Add `getSortByIndex` after `EmptyWS`, and add `import XMonad.Util.WorkspaceCompare` at the top.

**Reply by gmayerw** [reply]
```
XMonad will use ghc to recompile, because "/home/wgm/.xmonad/build" does not exist.
Error detected while loading xmonad configuration file: /home/wgm/.xmonad/xmonad.hs

xmonad.hs:62:106: error:
    • No instance for (Semigroup
                         (W.StackSet
                            WorkspaceId (Layout Window) Window ScreenId ScreenDetail))
        arising from a use of ‘<>’
    • In the first argument of ‘windows’, namely
        ‘(W.view ws <> W.shiftWin ws w)’
      In the expression: windows (W.view ws <> W.shiftWin ws w)
      In the first argument of ‘withFocused’, namely
        ‘(\ w -> windows (W.view ws <> W.shiftWin ws w))’
   |
62 |         , ("M-C-<Space>", doTo Prev EmptyWS getSortByIndex (\ws -> withFocused (\w -> windows (W.view ws <> W.shiftWin ws w))))
   |                 ^^

Please check the file for errors.

xmonad: xmessage: executeFile: does not exist (No such file or directory)


```

**Reply by geekosaur** [reply]
Oh, dear, so many 0.15 issues… change `<>` to `<+>`.

**Reply by gmayerw** [reply]
```
XMonad will use ghc to recompile, because "/home/wgm/.xmonad/build" does not exist.
Error detected while loading xmonad configuration file: /home/wgm/.xmonad/xmonad.hs

xmonad.hs:62:106: error:
    • No instance for (Monoid
                         (W.StackSet
                            WorkspaceId (Layout Window) Window ScreenId ScreenDetail))
        arising from a use of ‘<+>’
    • In the first argument of ‘windows’, namely
        ‘(W.view ws <+> W.shiftWin ws w)’
      In the expression: windows (W.view ws <+> W.shiftWin ws w)
      In the first argument of ‘withFocused’, namely
        ‘(\ w -> windows (W.view ws <+> W.shiftWin ws w))’
   |
62 |         , ("M-C-<Space>", doTo Prev EmptyWS getSortByIndex (\ws -> withFocused (\w -> windows (W.view ws <+> W.shiftWin ws w))))
   |                 ^^^

Please check the file for errors.

xmonad: xmessage: executeFile: does not exist (No such file or directory)


```

**Reply by geekosaur** [reply]
Hrm, that seems wrong.

Wait, `WindowSet` vs. `WindowSet -> WindowSet`. Lemme dig.

**Reply by geekosaur** [reply]
Right, I'm confusing `windows` with a `manageHook`. Replace `<+>` with `.`.

**Reply by gmayerw** [reply]
It compiles and is functioning now but it goes back one workspace. Anyone have ideas on how to get the original stated functionality? Seems we are getting closer.

**Reply by geekosaur** [reply]
It should go back to the previous empty workspace from the one you are currently on. As I stated [earlier](https://github.com/xmonad/xmonad/discussions/369#discussioncomment-2005698), it isn't really possible to have it go forward from the first workspace.

**Comment by gmayerw** [comment]
Is there anyway I can upgrade to the latest version? I don't have any
version constraints set and I am running pacman package manager on Arch
Linux

**Reply by geekosaur** [reply]
My understanding is that there is no longer a maintainer for the package on Arch, so you would have to install it via cabal or stack instead.

**Comment by gmayerw** [comment]
Basically it works, but it is taking me to the last workspace available and empty instead of the first one

**Reply by geekosaur** [reply]
Right, that's what I warned you it would do. There's no good way to do the other without some weird corner cases (like never putting anything on workspace 1).

**Comment by gmayerw** [comment]
Ok thank you I changed the `Prev` to `Next` and that functionality is close to what I am looking for. Thanks!

**Comment by gmayerw** [comment]
I was thinking that this may work if I use the existing code:
**Figure 1**
```
        , ("M-C-<Space>", doTo Next EmptyWS getSortByIndex (\ws -> withFocused (\w -> windows (W.view ws . W.shiftWin ws w))) >> asks (layoutHook . config) >>= setLayout)
```

But added to the algorithm. If before it does the above step, it always moves the window to workspace 9, it will always go to the next available window that is closest to 1 or is 1. For example, let's say that the target window is on workspace 3. I have a window open on workspace 2. There is also a window open on workspace 1. And one open on workspace 9. If I closed the one on 2, and run the code, it would ideally send the window to workspace 9 _(This is the part I need help with)_ who is already populated. Once there, the code from figure 1 would skip 1 because it is populated. 2 is not populated now since I closed it. The window will land on 2. Thus it started on 3, a workspace greater than 2 but was able to land back a window, keeping it as close as possible to 1. It would be a nice algorithm if I can get some help on how to send the window to 9, then run figure 1's code.

**Reply by gmayerw** [reply]
Also it is ok if 9 is empty in regards to corner cases. I'd like it to default there.

**Reply by gmayerw** [reply]
Checking on this

**Reply by geekosaur** [reply]
Sorry, I've been chewing on this a bit trying to figure a better way to do it. (Like I have a better way to do the earlier one that avoids calling `windows` for each individual window.) But I'm not thinking of one.

You would need to put before the `doTo` something like
```
withFocused (windows . W.shiftWin "9") >>
```
(so the `doTo` follows the `>>`). The problem is, after that it's not focused any more. Hm.

```
withFocused (\w -> windows (W.shiftWin "9" w) >>
                   doTo Next emptyWS getSortByIndex (\ws -> windows (W.view ws . W.shiftWin ws w))) >>
asks (layoutHook . config) >>= setLayout
```
This at least typechecks here.

**Comment by gmayerw** [comment]
![1648101674](https://user-images.githubusercontent.com/73801421/159852517-3b6d6a1d-304e-4752-81ea-4bd78ba81f26.png)

**Reply by gmayerw** [reply]
Changed `emptyWS` to `EmptyWS`.

After restarting, looks like there is still a problem: it functions the same as I had before.

**Comment by geekosaur** [comment]
Sorry, forgot you're still on 0.15, and 0.17.0 throws deprecation warnings with `EmptyWS`.

And I was afraid of it not working; I suspect it's got to shift you to workspace 9 for the `doTo` to operate correctly. It's a minor change but may be annoying in use.
```
withFocused (\w -> windows (W.view "9" . W.shiftWin "9" w) >>
                   doTo Next EmptyWS getSortByIndex (\ws -> windows (W.view ws . W.shiftWin ws w))) >>
asks (layoutHook . config) >>= setLayout
```

**Comment by gmayerw** [comment]
Wow it works beautifully! Thank you for your help! Great work!! It does EXACTLY what I was looking for! You really understood the problem. Thank you
